### PR TITLE
#2988 Exception when memory hint is too big

### PR DESCRIPTION
**Pull request by mauricio1802**
#2988 
When a big value as the used in the issue description is set as ```MemoryHint```, the ```Nethermind.Trie.MemmoryAllowance.TrieNodeCacheCount``` value will overflow because it is casted to ```int```, causing that the ```maxCapacity``` of the ```LruCache``` could be negative therefore the first time ```LruCache.Set``` is called it will try to remove the first element that doesn't exist.

This PR add the following changes: 

- Cast to ```int``` after divison when computing ```Nethermind.Trie.MemmoryAllowance.TrieNodeCacheCount```
- Thow ```ArgumentoOutOfRangeException``` when LruCache constructor is called with a ```maxCapacity``` lower than 1.
- Add test to check  ```maxCapacity``` argument validation.
- Add test to check big values as HintMemory.

### CODE REVIEW

File: src/Nethermind/Nethermind.Core.Test/Caching/LruCacheTests.cs [hunk 1]
```diff
@@ -172,5 +172,17 @@ public void Delete_keeps_internal_structure()
 
             count.Should().Be(itemsToKeep);
         }
+
+        [Test]
+        public void Wrong_capacity_number_at_constructor()
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Core.Test/Caching/LruCacheTests.cs:177]: so sweet to see a test as the first thing

File: src/Nethermind/Nethermind.Core.Test/Caching/LruCacheTests.cs [hunk 2]
```diff
@@ -172,5 +172,17 @@ public void Delete_keeps_internal_structure()
...
+        {
+            int maxCapacity = 0;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Core.Test/Caching/LruCacheTests.cs:179]: but we are missing the other case test -> a regression test would be good that shows that previously above int numbers did not work?

File: src/Nethermind/Nethermind.Trie/MemoryAllowance.cs [hunk 3]
```diff
@@ -22,6 +22,6 @@ public static class MemoryAllowance
     {
         public static long TrieNodeCacheMemory { get; set; } = 128.MB();
 
-        public static int TrieNodeCacheCount => (int)TrieNodeCacheMemory / PatriciaTree.OneNodeAvgMemoryEstimate;
+        public static long TrieNodeCacheCount => TrieNodeCacheMemory / PatriciaTree.OneNodeAvgMemoryEstimate;
```
Comment (tkstanczak) [inline src/Nethermind/Nethermind.Trie/MemoryAllowance.cs:25]: should not the fix be: (int)(TrieNodeCacheMemory / PatriciaTree.OneNodeAvgMemoryEstimate)
I would like to see the test showing how it was broken before and why it is fixed now

anyway, love the fact that a simple thing was fixed fast

File: src/Nethermind/Nethermind.Core.Test/Caching/LruCacheTests.cs (hunk 2 above)
Comment (mauricio1802) [reply src/Nethermind/Nethermind.Core.Test/Caching/LruCacheTests.cs:179]: Sure, you are right, I will add it ASAP

File: src/Nethermind/Nethermind.Trie/MemoryAllowance.cs (hunk 3 above)
Comment (mauricio1802) [reply src/Nethermind/Nethermind.Trie/MemoryAllowance.cs:25]: I see what you mean, but I thought that changing the type to long will allow bigger values, if (TrieNodeCacheMemory/PatriciaTree.OneNodeAvgMemoryEstimate) overflows int we would still have the error. What do you think would be better?

**Review by tkstanczak — APPROVED** [review]

**Review by matilote — APPROVED** [review]

**Pull request merged**